### PR TITLE
SE-0309: Unlock existential types for all protocols

### PR DESCRIPTION
**Pull request by AnthonyLatsis**

### CODE REVIEW

File: userdocs/diagnostics/existential-member-access-limitations.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Using Protocol Members with References to `Self` or `Self`-rooted Associated Types
+
+A protocol may be interfaced with via a conformance constraint on a generic parameter, an opaque result type, or via the protocol type itself:
```
Comment (harlanhaskins) [inline userdocs/diagnostics/existential-member-access-limitations.md:3]:
```suggestion
Protocol requirements may be accessed via a conformance constraint on a generic parameter, an opaque result type, or via the protocol type itself:
```

File: userdocs/diagnostics/existential-member-access-limitations.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  shape.matches(shape)
+}
+
+func foo(_ arg: Identifiable) {
+  // error: member 'id' cannot be used on value of protocol type 'Identifiable'; use a generic constraint instead
+  _ = arg.id
+}
+```
+
+An expection to this limitation are members that contain `Self` only in [covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) position (such as a method result type), where  `Self` can be safely substituted with the protocol or protocol composition type used to access the member — a representable supertype. On the other hand, resorting to this ploy in contravariant parameter type position, like allowing one to pass a type-erased value to a method that accepts `Self`, is not type-safe and would expose the opportunity to pass in an argument of non-matching type.
```
Comment (harlanhaskins) [inline userdocs/diagnostics/existential-member-access-limitations.md:42]:
```suggestion
An exception to this limitation are members that contain `Self` only in [covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) position (such as a method result type), where  `Self` can be safely substituted with the protocol or protocol composition type used to access the member — a representable supertype. On the other hand, resorting to this ploy in contravariant parameter type position, like allowing one to pass a type-erased value to a method that accepts `Self`, is not type-safe and would expose the opportunity to pass in an argument of non-matching type.
```

**Comment by AnthonyLatsis** [comment]
@swift-ci please test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test Linux

**Review by ktoso — APPROVED** [review]
LGTM and let's hunt down the issue introduced in https://github.com/apple/swift/pull/34140 separately. (more info in: https://github.com/apple/swift/pull/33767#issuecomment-908575354 )

File: lib/Serialization/ModuleFormat.h [hunk 3]
```diff
@@ -1257,7 +1257,6 @@ namespace decls_block {
     BCFixed<1>,             // implicit flag
     BCFixed<1>,             // class-bounded?
     BCFixed<1>,             // objc?
-    BCFixed<1>,             // existential-type-supported?
```
Comment (CodaFi) [inline lib/Serialization/ModuleFormat.h:1260]: Don't forget to bump the format version.

Comment (AnthonyLatsis) [reply lib/Serialization/ModuleFormat.h:1260]: Oops. Does it still need to be (has been incremented 6 times since this landed)?

**Pull request merged**